### PR TITLE
[BEAM-1909] Fix BigQuery read transform fails for DirectRunner when querying non-US regions

### PR DESCRIPTION
**Pull request by DannyLee12**
Basic
---
Added a fix to the Issue mentioned above.

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Format the pull request title like [BEAM-XXX] Fixes bug in ApproximateQuantiles, where you replace BEAM-XXX with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
- [x]  If this contribution is large, please file an Apache Individual Contributor License Agreement.

Includes
---
- Add regex to get projectid, datasetid and table id from both legacy and standard sql when project is provided
- Add regex to get datasetId and tableId from both legacy and standardsql when project is NOT provided.
- Add ability to use project attribute that already existed on BigQuerySource class
- If project is not provided at all, do not break the existing distributions. Therefore, nothing happens and the tables are created in location=None. This defaults back to US on directRunner but *should* work when running in dataflow.

Cavets
---
- I have not updated the unit tests, if required, please point me in the right direction.

Confirmation
---
```python
...beam.io.BigQuerySource(
                      query="""SELECT VehicleId
                               FROM [cartrack.vehicle_info]""",
                      use_standard_sql=True, project='sa-taxi-edw'))
```
```
Dataset sa-taxi-edw:temp_dataset_297f3d8de4364961bf63819985e690c8 does not exist so we will create it as temporary with location=EU
```

```python
...beam.io.BigQuerySource(
                      query="""SELECT VehicleId
                               FROM `cartrack.vehicle_info`""",
                      use_standard_sql=False, project='sa-taxi-edw'))
```
```
WARNING:root:Dataset sa-taxi-edw:temp_dataset_6cf4c45c5ff54b69a60c0851aba33398 does not exist so we will create it as temporary with location=EU
```
```python
...beam.io.BigQuerySource(
                      query="""SELECT VehicleId
                               FROM `sa-taxi-edw.cartrack.vehicle_info`""",
                      use_standard_sql=True))
```
```
WARNING:root:Dataset sa-taxi-edw:temp_dataset_16f490456e744fa78278b2cd95fe3c5b does not exist so we will create it as temporary with location=EU
```

```python
...beam.io.BigQuerySource(
                      query="""SELECT VehicleId
                               FROM [sa-taxi-edw:cartrack.vehicle_info]""",
                      use_standard_sql=False))
```
```
WARNING:root:Dataset sa-taxi-edw:temp_dataset_2cab90333b8049aab4d2adbae5fc65e0 does not exist so we will create it as temporary with location=EU
```

In the instance that the project is provided neither in the query nor as an argument, the location is set to None. In this instance, the location defaults to US when using the DirectRunner.

```
WARNING:root:Dataset sa-taxi-edw:temp_dataset_8def137665b94f1fb128a23edcf0aa19 does not exist so we will create it as temporary with location=None

```

### CODE REVIEW

**Comment by DannyLee12** [comment]
Can anyone tell me why these commits are failing? I didn't touch any java code. I only edited one file.

**Comment by DannyLee12** [comment]
```
WARNING:root:Dataset sa-taxi-edw:temp_dataset_9b7e2af67b4841d3970f89b238fedac9 does not exist so we will create it as temporary with location=None
```

Still getting this issue when I try and run this in Dataflow. Please review and merge when you get a chance?

**Comment by angoenka** [comment]
R: @udim 
Can you please take a look. 
Thanks!

**Comment by apilloud** [comment]
run python precommit

**Comment by apilloud** [comment]
ping @udim

File: sdks/python/apache_beam/io/gcp/bigquery.py [hunk 1]
```diff
@@ -641,19 +642,76 @@ def __init__(self, source, test_bigquery_client=None, use_legacy_sql=True,
...
+    if pm:
+      return pm.group(1), pm.group(2), pm.group(3)
+    else:
+      npm = re.search(not_project_regex, self.source.query)
+      if npm:
+        if self.source.project:
+          return self.source.project, npm.group(1), npm.group(2)
+    return (None for _ in range(3))  # No matches
+
+  def _parse_query(self):
```
Comment (udim) [inline sdks/python/apache_beam/io/gcp/bigquery.py]: We don't want to parse BigQuery SQL in Beam, as having our own parser would take a lot of effort to maintain.
Instead, the preferred way to discover tables in a query is to run it in dry-run mode and look at the list of referenced tables, as mentioned in the bug:
https://issues.apache.org/jira/browse/BEAM-1909?focusedCommentId=16020138&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16020138

Dry run option:
https://github.com/apache/beam/blob/e278e57077d8ab6459a9833eccd2ef20dc2faeae/sdks/python/apache_beam/io/gcp/bigquery.py#L806

REST API reference for referencedTables:
https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs#statistics.query.referencedTables

**Pull request closed without merging**